### DISCUSSION TITLE
To get bam files

### DISCUSSION THREAD

**Discussion by AnjaliC4** [GreenleafLab/ArchR#1346] [Questions / Documentation]
Hi @rcorces,

To get bam files per cell-type from ArchR - I would have to use getFragmentsFromProject() with specific cell-names, convert GRanges object to bed format, subtract -1 from start coordinate (as bed files are zero-indexed), and then use bedtobam function from bedtools to obtain final bam files. Does that sound right to you? 

Also, if I have missed a simpler way to get generate them from ArchR - please let me know.

Thanks

**Comment by rcorces** [comment]
Assuming you have standard 10x data, I would use the `cellNames` from ArchR to get the `RG` tag and then use samtools to extract the reads from the original bam files.
https://bioinformatics.stackexchange.com/questions/10559/extract-reads-from-bam-files-by-their-rg

We've discussed making this more straightforward with ArchR but to my knowledge Rsamtools cannot directly export bam files.

**Reply by AnjaliC4** [reply]
This may be difficult when there are multiple subjects as if you have multiple subjects, one would have to extract reads individually from them and change barcodes to make sure duplicate barcodes are treated properly -which is not straightforward.

**Reply by Zepeng-Mu** [reply]
I've been using `sinto`. You can export a file with barcodes and cell labels, and then extracts bam reads for each label. But it's true that you need to this for each 10x library separately, and this can be annoying if your ArchR project contains many libraries. Maybe it's easier to run `cell ranger aggr` first to combine 10x bam files and then do downstream analysis?

**Reply by AnjaliC4** [reply]
yeah that is a good idea to combine them but unfortunately cell ranger -aggr does not produce an aggregated bam file - only fragment file and count matrices.

**Reply by rcorces** [reply]
I dont *think* this would be that hard to add to ArchR, at least to get SAM output. Time is quite limited but I can add it to the growing list or review a pull request for a new function.

**Reply by AnjaliC4** [reply]
@rcorces can you please point me to the source code of the function where this could be added to produce SAM output?

**Reply by rcorces** [reply]
it would need to be a totally new function. Probably in https://github.com/GreenleafLab/ArchR/blob/release_1.0.2/R/ArrowRead.R

**Comment by AnjaliC4** [comment]
Hi @rcorces, I looked into this further and it maybe a bit out of my expertise to add a new function quickly. Since, many tools need cell-type bams to run, it would be truly worthwhile and helpful if this can be implemented. Can you please tell me a timeline by which you can expect this to be implemented?

Even if you can provide me with some basic code of how to do this most efficiently within ArchR- that would be wonderful. Thanks!

**Reply by rcorces** [reply]
sorry. cant provide a timeline.

**Comment by AnjaliC4** [comment]
Hi @rcorces, is there an update on the implementation of this in ArchR? It seems simple/possible to implement this within the file GroupExport.R (as this exports groupBW)
https://rdrr.io/github/GreenleafLab/ArchR/src/R/GroupExport.R

(but unfortunately its very very difficult to do outside ArchR when you have many samples (problems include: barcode name collision, filtering specific cells from many files, merging bams properly etc) 

It would be very helpful to have this function, as very often one needs group bam files for many downstream analyses. 
Thanks again!

**Reply by rcorces** [reply]
I dont have any update on this. its on my list but I do not anticipate this being added soon.
